packages/block-editor/src/components/block-list/index.js

```diff
@@ -202,122 +169,105 @@ class BlockList extends Component {
-								// otherwise there might be a small delay to trigger the animation.
-								animateOnChange={ index }
-								enableAnimation={ enableAnimation }
-							/>
-						</BlockAsyncModeProvider>
-					);
-				} ) }
-
-				<BlockListAppender
-					rootClientId={ rootClientId }
-					renderAppender={ renderAppender }
-				/>
-			</div>
-		);
-	}
-}
+	/**
+	 * Handles a mouseup event to end the current mouse multi-selection.
+	 */
+	const onSelectionEnd = () => {
+		// Equivalent to attaching the listener once.
+		window.removeEventListener( 'mouseup', onSelectionEnd );
+		// The browser selection won't have updated yet at this point, so wait
+		// until the next animation frame to get the browser selection.
+		rafId.current = window.requestAnimationFrame( setSelection );
+	};
 
-export default compose( [
-	// This component needs to always be synchronous
-	// as it's the one changing the async mode
-	// depending on the block selection.
-	forceSyncUpdates,
-	withSelect( ( select, ownProps ) => {
-		const {
-			getBlockOrder,
-			isSelectionEnabled,
-			isMultiSelecting,
-			getMultiSelectedBlocksStartClientId,
-			getMultiSelectedBlocksEndClientId,
-			getSelectedBlockClientId,
-			getMultiSelectedBlockClientIds,
-			hasMultiSelection,
-			getGlobalBlockCount,
-			isTyping,
-		} = select( 'core/block-editor' );
+	// Only clean up when unmounting, these are added and cleaned up elsewhere.
+	useEffect( () => () => {
+		window.removeEventListener( 'mouseup', onSelectionEnd );
+		window.cancelAnimationFrame( rafId.current );
+	}, [] );
 
-		const { rootClientId } = ownProps;
+	/**
+	 * Binds event handlers to the document for tracking a pending multi-select
+	 * in response to a mousedown event occurring in a rendered block.
+	 */
+	const onSelectionStart = () => {
+		if ( ! isSelectionEnabled ) {
+			return;
+		}
 
-		return {
-			blockClientIds: getBlockOrder( rootClientId ),
-			selectionStart: getMultiSelectedBlocksStartClientId(),
-			selectionEnd: getMultiSelectedBlocksEndClientId(),
-			isSelectionEnabled: isSelectionEnabled(),
-			isMultiSelecting: isMultiSelecting(),
-			selectedBlockClientId: getSelectedBlockClientId(),
-			multiSelectedBlockClientIds: getMultiSelectedBlockClientIds(),
-			hasMultiSelection: hasMultiSelection(),
-			enableAnimation: (
-				! isTyping() &&
-				getGlobalBlockCount() <= BLOCK_ANIMATION_THRESHOLD
-			),
-		};
-	} ),
-	withDispatch( ( dispatch ) => {
-		const {
-			startMultiSelect,
-			stopMultiSelect,
-			multiSelect,
-		} = dispatch( 'core/block-editor' );
+		startMultiSelect();
 
-		return {
-			onStartMultiSelect: startMultiSelect,
-			onStopMultiSelect: stopMultiSelect,
-			onMultiSelect: multiSelect,
-		};
-	} ),
-] )( BlockList );
+		// `onSelectionStart` is called after `mousedown` and `mouseleave`
+		// (from a block). The selection ends when `mouseup` happens anywhere
+		// in the window.
+		window.addEventListener( 'mouseup', onSelectionEnd );
```

Comment:
I think personally having this listened and its cleanup in the same effect is better in terms of clarity. Do you expect any noticeable impact on performance? The onSelectionEnd could check whether a multi-selection is in progress.

Reply:
It will depend on the type of content. If it's just a list of paragraphs, then no. If the content has many different block lists, like full site editing with complex content, then maybe? I think we should prevent any unnecessary listeners where possible. I don't think we should _not_ do it because it's an uncommon pattern. If it's unclear, I should comment better. :)